core-server/src/main/java/org/glassfish/jersey/server/model/RuntimeResource.java

```diff
@@ -128,7 +128,16 @@ public RuntimeResource build(RuntimeResource parent) {
     public static final Comparator<RuntimeResource> COMPARATOR = new Comparator<RuntimeResource>() {
         @Override
         public int compare(RuntimeResource o1, RuntimeResource o2) {
-            return PathPattern.COMPARATOR.compare(o1.getPathPattern(), o2.getPathPattern());
+            final int cmp = PathPattern.COMPARATOR.compare(o1.getPathPattern(), o2.getPathPattern());
+            int locatorCmp = 0;
+            return cmp != 0
+                    ? cmp
+                    // quaternary key sorting those derived from
+                    // sub-resource methods ahead of those derived from sub-resource locators
+                    : (locatorCmp = o1.resourceLocators.size() - o2.resourceLocators.size()) != 0
+                        ? locatorCmp
+                        // compare the regexes then
+                        : o2.regex.compareTo(o1.regex);
```

Comment:
Nested ternary operators create quite unreadable code. I would refactor this to use an if statement instead.
```
final int cmp = PathPattern.COMPARATOR.compare(o1.getPathPattern(), o2.getPathPattern());
        if (cmp == 0) {
            final int locatorCmp = Integer.compare(o1.resourceLocators.size(), o2.resourceLocators.size());
            // quaternary key sorting those derived from
            // sub-resource methods ahead of those derived from sub-resource locators

            // compare the regexes if still equal
            return (locatorCmp == 0) ? o2.regex.compareTo(o1.regex) : locatorCmp;
        } else {
            return cmp;
        }
```